xdsl/dialects/csl/csl_wrapper.py

```diff
@@ -0,0 +1,308 @@
+            ops.append(op)
+            fields.append(StringAttr(field))
+
+        super().__init__(
+            operands=[ops],
+            properties={
+                "module": StringAttr(module),
+                "fields": ArrayAttr(fields),
+            },
+            result_types=[csl.ImportedModuleType()],
+        )
+
+    def verify_(self) -> None:
+        if len(self.fields) != len(self.ops):
+            raise ValueError("Number of fields does not match number of operands")
+
+
+@irdl_op_definition
+class ModuleOp(IRDLOperation):
+    """
+    Wrapper class that manages initialisation of layout and program module.
+
+    Specified properties will be lowered to module params (`csl.param`). As such, all properties
+    are passed as BlockArgs to the `layout_module` and `program_module` region. The order in which
+    properties are specified is important and must match the order of the block args.
+
+    Additionally, any value yielded by the `layout_module` is passed as a block arg (lowered to `csl.param`)
+    to the program module. Order is important here as well. The layout module's `yield` op should be lowered to
+    `@set_tile_code`, while the program module's `yield` op should be discarded.
+
+    The layout module has two additional block args `x` and `y` as part of the `@set_tile_code` loop nest.
+    Operations using these args need to be lowered to the correct place in the loop nest.
+    """
+
+    name = "csl_wrapper.module"
+
+    width = prop_def(IntegerAttr)
+    height = prop_def(IntegerAttr)
+    params = prop_def(DictionaryAttr)
+
+    layout_module = region_def("single_block")
+    program_module = region_def("single_block")
+
+    @staticmethod
+    def from_properties(properties: dict[str, Attribute]):
+        layout_module = Region(Block())
+        program_module = Region(Block())
+        x = layout_module.block.insert_arg(IntegerType(16), 0)
+        y = layout_module.block.insert_arg(IntegerType(16), 1)
+        x.name_hint = "x"
+        y.name_hint = "y"
+
+        for name, value in properties.items():
+            if not isa(value, IntegerAttr[IntegerType]):
+                raise ValueError("Can only create module from IntegerAttr properties")
+            l_arg = layout_module.block.insert_arg(
+                value.type, len(layout_module.block.args)
+            )
+            l_arg.name_hint = name
+            p_arg = program_module.block.insert_arg(
+                value.type, len(program_module.block.args)
+            )
+            p_arg.name_hint = name
+
+        props = {
+            "width": properties.pop("width"),
+            "height": properties.pop("height"),
+            "params": DictionaryAttr(properties),
+        }
+
+        return ModuleOp(properties=props, regions=[layout_module, program_module])
+
+    def update_program_block_args_from_layout(self):
+        """Update `program_module` BlockArguments by adding yield op fields"""
```

Comment:
> I would like to see an explanation that the yield op parameters are additional inputs to the program code, as this is not obvious from the outside.

In the context of this method this is probably the only thing this sentence intends to say. It's explained in more detail in the doc string of the class.

> Also, would it be possible for this method to be idempotent? I am myself not 100% certain, but I would like to see it be the case.

What do you mean by this exactly? There's currently an assert set up to fail when invoking the method more than once. Alternatively, it could return silently, or else it could attempt to do some kind of recovery if things are not in the expected state.